experimental/sql/src/projection_optimizations.py

```diff
@@ -0,0 +1,191 @@
+# projections. Currently, there are three rewriters. The first one introduces a
+# projection before any operator, the second one fuses subsequent projections,
+# and the third one removes projetions that are just identity mappings.
+
+
+@dataclass
+class ProjectionOptimizer(RewritePattern):
+
+  def flatten(self, l: list[list[str]]) -> list[str]:
+    return [elem for sublist in l for elem in sublist]
+
+  def find_cols_in_expr(self, op: RelAlg.Expression) -> list[str]:
+    """
+    Returns the list of cols that are used in this expression or its children.
+    """
+    if isinstance(op, RelAlg.Column):
+      return [op.col_name.data]
+    else:
+      return self.flatten([self.find_cols_in_expr(r.op) for r in op.regions])
+
+  def find_cols_upstream(self, op: RelAlg.Operator) -> list[str]:
+    """
+    Returns the list of cols that are used in this operation or its parents. If
+    the list contains None, a ModuleOp (so the output) was hit, showing that all
+    columns need to be kept.
+    """
+    if isinstance(op, ModuleOp):
+      return [None]
+    if isinstance(op, RelAlg.Select):
+      return self.flatten([
+          self.find_cols_in_expr(o) for o in op.predicates.ops
+      ]) + self.find_cols_upstream(op.parent_op())
+    if isinstance(op, RelAlg.Project):
+      return self.flatten(
+          [self.find_cols_in_expr(o) for o in op.projections.ops])
+    if isinstance(op, RelAlg.Aggregate):
+      return [s.data for s in op.col_names.data]
+
+  def find_schema(self, op: RelAlg.Operator) -> list[str]:
+    """
+    Returns the list of cols that are the input schema of this operation.
+    """
+    if isinstance(op, RelAlg.Table):
+      return [o.elt_name.data for o in op.schema.ops]
+    if isinstance(op, RelAlg.Select):
+      return self.find_schema(op.input.op)
+    if isinstance(op, RelAlg.Project):
+      return [s.data for s in op.names.data]
+    if isinstance(op, RelAlg.Aggregate):
+      return [s.data for s in op.col_names.data]
+
+
+@dataclass
+class ProjectionInference(ProjectionOptimizer):
+
+  # This rewriter infers columns in front of all Operators (but projections and
+  # tables) that map away all columns that aren't needed by this Operator or
+  # further up the tree.
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelAlg.Operator, rewriter: PatternRewriter):
+    if isinstance(op, RelAlg.Project) or isinstance(
+        op, RelAlg.Table) or isinstance(op, RelAlg.SchemaElement):
+      return
+    cols = list(dict.fromkeys(self.find_cols_upstream(op)))
+    if cols.__contains__(None):
+      return
+
+    new_proj = RelAlg.Project.get(
+        rewriter.move_region_contents_to_new_regions(op.input),
+        Region.from_operation_list([RelAlg.Column.get(s) for s in cols]),
+        ArrayAttr.from_list([StringAttr.from_str(s) for s in cols]))
+
+    # TODO: I mainly need this pattern match to call the proper builder. This
+    # could maybe be done nicer.
+    if isinstance(op, RelAlg.Select):
```

Comment:
I wish that this wouldn't need a type switch here. Is there no generic way to get all operator inputs here and replace them with the new input?